src/ai/backend/agent/observer/host_port.py

```diff
@@ -0,0 +1,65 @@
+import logging
+from collections import defaultdict
+from typing import TYPE_CHECKING, Optional, override
+
+from ai.backend.common.observer.types import AbstractObserver
+from ai.backend.common.types import ContainerStatus
+from ai.backend.logging.utils import BraceStyleAdapter
+
+if TYPE_CHECKING:
+    from ..agent import AbstractAgent
+
+
+log = BraceStyleAdapter(logging.getLogger(__spec__.name))
+
+
+PORT_USAGE_THRESHOLD = 2
+
+
+class HostPortObserver(AbstractObserver):
+    def __init__(
+        self,
+        agent: "AbstractAgent",
+    ) -> None:
+        self._agent = agent
+        self._real_used_candidate_ports: defaultdict[int, int] = defaultdict(int)
+
+    @property
+    @override
+    def name(self) -> str:
+        return "agent_host_port"
```

Comment:
It looks better to make it with a classmethod, but there's no need to apply it right away.